### PR TITLE
Merge BSpline Surfaces

### PR DESCRIPTION
**Pull request by sbendjebla**
* **Please check if the PR fulfills these requirements**
- [ ] The commit messages follows our guidelines (explicit, in english)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] CHANGELOG.md has been updated

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [x] Bug fix:
- [ ] New features:
- [ ] Documentation
- [ ] Optimization

* **Other information**:
_I improved the methods for merging bspline surface/faces._

### CODE REVIEW

File: volmdlr/faces.py [hunk 1]
```diff
@@ -5747,104 +5652,200 @@ def adjacent_direction_uv(nearest_start1, nearest_start2):
...
+            pt = p.nearest_point(points2)
+            ind.append(points2.index(pt))
+            dis.append(p.point_distance(pt))
+
+        dis_sorted = sorted(dis)
+
+        shared = []
+        for k, p1 in enumerate(contour1.primitives):
+            if dis_sorted[0] == dis_sorted[1]:
+                indices = npy.where(npy.array(dis) == dis_sorted[0])[0]
```
Comment (AxelRinghausen) [inline volmdlr/faces.py:5736]: I don't really understand what npy.where does. Can you enlighten me ?

File: volmdlr/faces.py [hunk 2]
```diff
@@ -5747,104 +5652,200 @@ def adjacent_direction_uv(nearest_start1, nearest_start2):
...
-        
-        size_u, size_v, degree_u, degree_v = 20, 10, max(bsplines[0].degree_u, bsplines[1].degree_u), max(bsplines[0].degree_v, bsplines[1].degree_v)
-        
-        merged_surface = volmdlr.faces.BSplineSurface3D.points_fitting_into_bspline_surface(points3d, size_u, size_v, degree_u, degree_v)
-                
-        return merged_surface
+        merged_surface = self.surface3d.merge_with(other_bspline_face3d.surface3d)
+        merged_face = merged_surface.rectangular_cut(0,1,0,1)
+
+        return merged_face
```
Comment (AxelRinghausen) [inline volmdlr/faces.py:5848]: Shouldn't you cut the `merged_surface` with `self.surface2d.merge_with(other_bspline_face3d.surface2d)`  instead of a `rectangular_cut` ?

File: volmdlr/wires.py [hunk 3]
```diff
@@ -3878,27 +3879,36 @@ def tessellation_points(self, resolution=20):
...
             angle = (theta1 + theta2) / 2
 
-        p_3 = volmdlr.Point3D(self.major_axis * math.cos(angle),
-                              self.minor_axis * math.sin(angle), 0)
-        p3 = frame.old_coordinates(p_3)
+        # p_3 = volmdlr.Point3D(self.major_axis * math.cos(angle),
+        #                       self.minor_axis * math.sin(angle), 0)
+        # p3 = frame.old_coordinates(p_3)
+        p3 = frame.old_coordinates(volmdlr.Point3D(self.major_axis * math.cos(angle),
+                                                   self.minor_axis * math.sin(angle), 0))
```
Comment (AxelRinghausen) [inline volmdlr/wires.py:3911]: Why these changes ? (Intermediate variables have been removed)

**Review by AxelRinghausen — CHANGES_REQUESTED** [review]

File: volmdlr/faces.py [hunk 4]
```diff
@@ -3036,75 +2865,81 @@ def points_approximate_into_bspline_surface(cls, points_3d, size_u, size_v, degr
...
 
-            points_3d = faces[0].surface3d.grid3d(points_x, points_y,
-                                                  faces[0].surface2d.outer_contour.bounding_rectangle()[0],
-                                                  faces[0].surface2d.outer_contour.bounding_rectangle()[1],
-                                                  faces[0].surface2d.outer_contour.bounding_rectangle()[2],
-                                                  faces[0].surface2d.outer_contour.bounding_rectangle()[3])
+        if len(cylindrical_faces) == 1:
 
-            return volmdlr.faces.BSplineSurface3D.points_fitting_into_bspline_surface(points_3d,points_x,points_x,degree_u,degree_v)
+            return volmdlr.faces.BSplineSurface3D.from_cylindrical_face(cylindrical_faces[0], degree_u, degree_v, 50, 50)
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply volmdlr/faces.py]: It's done!

File: volmdlr/faces.py [hunk 5]
```diff
@@ -3472,6 +3312,118 @@ def f(X):
...
+                    xmax.append(1)
+                    ymin.append(0.01)
+                    ymax.append(1)
+
+                xmin.append(0)
+                xmax.append(1)
+                ymin.append(0)
+                ymax.append(1)
+
+                return xmin, xmax, ymin, ymax
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply volmdlr/faces.py]: It's done!

File: volmdlr/faces.py [hunk 6]
```diff
@@ -5579,164 +5555,93 @@ def triangulation_lines(self, resolution=25):
...
                 grid2d_direction = [['+x','+y'], ['+x','+y']]
-    
+
+
             elif (v1 == 0 and v2 == 1):
                 corresponding_directions.append(('+v', '+v'))
                 grid2d_direction = [['+x','-y'], ['+x','-y']]
-    
+
             return corresponding_directions, grid2d_direction
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply volmdlr/faces.py]: It's done!

File: volmdlr/faces.py [hunk 7]
```diff
@@ -5747,104 +5652,200 @@ def adjacent_direction_uv(nearest_start1, nearest_start2):
...
                
             elif (v1 == 0 and u2 == 0):
                 corresponding_directions.append(('+v', '-u'))
                 grid2d_direction = [['-x','-y'], ['-y','+x']]
-    
+            
             return corresponding_directions, grid2d_direction
-    
+
         def adjacent_direction_vu(nearest_start1, nearest_start2):
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply volmdlr/faces.py]: It's done!

File: volmdlr/faces.py [hunk 8]
```diff
@@ -5579,164 +5555,93 @@ def triangulation_lines(self, resolution=25):
...
+
             elif (v1 == 0 and v2 == 1):
                 corresponding_directions.append(('+v', '+v'))
                 grid2d_direction = [['+x','-y'], ['+x','-y']]
-    
+
             return corresponding_directions, grid2d_direction
-                             
+
         def adjacent_direction_vv(nearest_start1, nearest_start2):
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply volmdlr/faces.py]: It's done!

File: volmdlr/faces.py [hunk 9]
```diff
@@ -3472,6 +3312,118 @@ def f(X):
...
+                    errors.append(s.error_with_point3d(bsplines[i].point2d_to_3d(center[i])))
+
+                bsplines_new[i] = surfaces[errors.index(min(errors))]
+
+            xmin, xmax, ymin, ymax = [], [], [], []
+            for i in range(0, len(bsplines_new)):
+                xmin.append(0)
+                xmax.append(1)
+                ymin.append(0)
+                ymax.append(1)
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply volmdlr/faces.py]: Thanks!

File: volmdlr/faces.py [hunk 10]
```diff
@@ -5747,104 +5652,200 @@ def adjacent_direction_uv(nearest_start1, nearest_start2):
...
+
+        contour1_2d = self.surface2d.outer_contour
+        contour2_2d = other_bspline_face3d.surface2d.outer_contour
+
+        points1 = []
+        for p1 in contour1.primitives:
+            points1.append(p1.start)
+        points2 = []
+        for p2 in contour2.primitives:
+            points2.append(p2.start)
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply volmdlr/faces.py]: Thanks!

File: volmdlr/wires.py (hunk 3 above)
Comment (sbendjebla) [reply volmdlr/wires.py:3911]: Finally, we keep these changes or not ?

File: volmdlr/faces.py (hunk 2 above)
Comment (sbendjebla) [reply volmdlr/faces.py:5875]: It's done !

File: volmdlr/wires.py (hunk 3 above)
Comment (AxelRinghausen) [reply volmdlr/wires.py:3911]: Yes, we can keep these changes.

File: volmdlr/faces.py [hunk 11]
```diff
@@ -5568,16 +5570,29 @@ def pair_with(self, other_bspline_face3d):
...
             corresponding_directions.append(('+' + adjacent_direction1, '-' + adjacent_direction2))
 
+        if adjacent_direction1 == 'u' and adjacent_direction2 == 'u':
+            corresponding_directions, grid2d_direction = self.adjacent_direction_uu(other_bspline_face3d, corresponding_directions)
+        elif adjacent_direction1 == 'v' and adjacent_direction2 == 'v':
+            corresponding_directions, grid2d_direction = self.adjacent_direction_vv(other_bspline_face3d, corresponding_directions)
+        elif adjacent_direction1 == 'u' and adjacent_direction2 == 'v':
+            corresponding_directions, grid2d_direction = self.adjacent_direction_uv(other_bspline_face3d, corresponding_directions)
+        elif adjacent_direction1 == 'v' and adjacent_direction2 == 'u':
+            corresponding_directions, grid2d_direction = self.adjacent_direction_vu(other_bspline_face3d, corresponding_directions)
```
Comment (AxelRinghausen) [inline volmdlr/faces.py:5588]: FYI: you can use the `getattr` keyword here.
```suggestion
        corresponding_directions, grid2d_direction = getattr(self, 'adjacent_direction_' + adjacent_direction1 + adjacent_direction2)(other_bspline_face3d, corresponding_directions)
```

**Review by AxelRinghausen — APPROVED** [review]
